Combine separate flags into a player state property

low-batt opened the pull request:
This commit will:
- Add a `PlayerState` struct that defines the mutually exclusive states a `PlayerCore` can be in
- Add a property to `PlaybackInfo` to hold a player's state
- Remove the `fileLoading`, `isStopping`, `isStopped`, `isShuttingDown` and `isShutdown` properties from `PlayerCore`
- Change the `isIdle`, i`sPaused`, `isPlaying` and `isSeeking` `PlaybackInfo` properties to be computed properties that check the state of the player
- Add an i`sActive` computed property to `PlaybackInfo` that checks if the player is in an active state
- Move remaining `MPV_EVENT_SHUTDOWN` processing from `MPVController` to `PlayerCore` and the main thread to avoid the need for locks
- Change handling of `MPV_EVENT_END_FILE` to check the player state instead of `MPV_END_FILE_REASON_STOP`
- Add an `isActive` guard to `PlaylistViewController.reloadData`

This makes the state the player is in explicit in one property. This also corrects some errors in the app termination sequence.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [ ] This implements/fixes issue #.

---

**Description:**

low-batt committed:
Combine separate flags into a player state property

This commit will:
- Add a PlayerState struct that defines the mutually exclusive states a
  PlayerCore can be in
- Add a property to PlaybackInfo to hold a player's state
- Remove the fileLoading, isStopping, isStopped, isShuttingDown and
  isShutdown properties from PlayerCore
- Change the isIdle, isPaused, isPlaying and isSeeking PlaybackInfo
  properties to be computed properties that check the state of the
  player

- Add an isActive computed property to PlaybackInfo that checks if the
  player is in an active state
- Move remaining MPV_EVENT_SHUTDOWN processing from MPVController to
  PlayerCore and the main thread to avoid the need for locks
- Change handling of MPV_EVENT_END_FILE to check the player state
  instead of MPV_END_FILE_REASON_STOP
- Add an isActive guard to PlaylistViewController.reloadData

This makes the state the player is in explicit in one property. This
also corrects some errors in the app termination sequence.

low-batt commented:
@lhc7000 @uiryuu This is the refactoring to eliminate the multiple player boolean flags.

low-batt commented:
You will see I left some extra logging in that will need to be changed to level `.verbose` before the beta release.

uiryuu commented on iina/PlayerState.swift:
```diff
@@ -0,0 +1,105 @@
+        if !result.isEmpty {
+          result += ","
+        }
+        result += state.description
```
I don't like the recursive implementation here, it's not very straightforward and prone to bugs. I think we can simply declare a dictionary of [Option, String] and use `filter` + `join` to form the description.

low-batt replied:
Will change the implementation to use a dictionary.

low-batt replied:
Switched to an `enum`, so this code was eliminated.

uiryuu commented on iina/PlaybackInfo.swift:
```diff
@@ -34,12 +34,67 @@ class PlaybackInfo {
+  var isIdle: Bool { state == .idle }
+  var isPaused: Bool { state == .paused }
+  var isPlaying: Bool { state == .playing }
+  var isSeeking: Bool { state == .seeking}
```
```suggestion
  var isSeeking: Bool { state == .seeking }
```

low-batt replied:
Good catch. As per the other comment these will be removed.

low-batt replied:
These computed properties have been removed.

uiryuu commented on iina/PlaybackInfo.swift:
```diff
@@ -34,12 +34,67 @@ class PlaybackInfo {
+  ///     transition from `.shuttingDown` to `.idle`.
+  var state: PlayerState {
+    get { _state }
+    set {
```
I suggest get rid of two state variables and merge the `set` logic here into `didSet`. `didSet` should be called directly after the value has been set and on the same thread, so there shouldn't be any thread issue. We can check if the change of state is valid, and revert back to `oldValue` if not valid.

Generally reverting back to the original valid in `didSet` is legal as per [Apple's documentation](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/properties/):

> Similarly, if you implement a didSet observer, it’s passed a constant parameter containing the old property value. You can name the parameter or use the default parameter name of oldValue. **If you assign a value to a property within its own didSet observer, the new value that you assign replaces the one that was just set.**

low-batt replied:
Will remove the `_state` variable and just use `didSet`.

uiryuu commented on iina/PlayerState.swift:
```diff
@@ -0,0 +1,105 @@
+import Foundation
+
+/// The mutually exclusive states a `PlayerCore` can be in.
+struct PlayerState: OptionSet, CustomStringConvertible {
```
Why `PlayerState` is implemented as an `OptionSet`? If I didn't miss anything, all the cases are mutually exclusive and we didn't use `PlayerState.insert()`. So a plain enum should do, no?

low-batt replied:
As per offline discussion I will add comment to explain why an `OptionSet` is being used. Then we will see what other reviewers think it is worthwhile using an `OptionSet` for performance.

low-batt replied:
Switched to an `enum`.

uiryuu commented on iina/AppDelegate.swift:
```diff
@@ -646,7 +646,7 @@ class AppDelegate: NSObject, NSApplicationDelegate, SPUUpdaterDelegate {
     // Instruct any players that are already stopped to start shutting down.
     for player in PlayerCore.playerCores {
-      if player.isStopped && !player.isShutdown {
+      if player.info.state == .stopped || player.info.state == .idle {
```
Here `player.info.state == .idle` is used; in other places `player.isIdle` is used. We should use only one of them in all places. 

For me, I don't see a particular advantage of using the computed properties (other than it will be a little bit shorter). And why we only have `isIdle, isPaused, isPlaying, isSeeking`, what about other states? So I suggest remove all the computed properties, and only use `player.info.state == __`. This style will remind us and new programmers explicitly that we are using `PlayerInfo.State` other than some unknown logic to compute the current state.

low-batt replied:
Will remove the computed properties and replace with direct checking of the state.

svobs replied:
> So I suggest remove all the computed properties, and only use player.info.state == __ 

In many places we need to check against a range of states. To ensure clarity, how about adding to `PlayerState`:

```Swift
  func isAtLeast(_ minState: PlayerState) -> Bool {
    return rawValue >= minState.rawValue
  }

  func isNotYet(_ state: PlayerState) -> Bool {
    return rawValue < state.rawValue
  }
```
So in many places, you'd have e.g., `guard player.info.state.isNotYet(.stopping) else { return }`.

svobs replied:
Responding to myself after reading more of the existing PR code...
> So in many places, you'd have e.g., `guard player.info.state.isNotYet(.stopping) else { return }`.

I found that in practice, the `[.playing, .paused, .seeking]` states are sort of off on their own in relation to the other states.

But writing `player.info.state.isAtLeast(.playing)` doesn't seem very intuitive.

But you don't have to minimize the number of states! You could add another state above `.playing`, called `.started`. That way you could write `player.info.state.isAtLeast(.started)`. But you'd never even have to set that one explicitly. Or you could use it to indicate that some kind of init is done.

low-batt replied:
I like this idea. I will see how it works out. Working on updating this PR today.

low-batt replied:
I liked the idea of these general methods, but I found there were only two cases where checking for multiple states, so instead I added `active` and `loaded` computed properties in the `enum`. They are checking the `rawValue` as suggested.

svobs replied:
> I liked the idea of these general methods, but I found there were only two cases where checking for multiple states, so instead I added active and loaded computed properties in the enum.

Sounds fine to me. But you should name them `isActive` & `isLoaded` which is the naming convention for booleans in most projects. You probably were following `mainWindow.loaded`, but that's a special case...it looks it was created as a substitute for `isLoaded` which already exists but has unwanted side effects.

low-batt replied:
I first coded them as `isActive` and `isLoaded` because I too am used to that convention. I renamed them because I didn't see that convention consistently used. Happy to change them back. Lets first see what @uiryuu has to say about IINA conventions.

low-batt replied:
I had a chat with @uiryuu about boolean flag naming conventions. The "is" prefix convention is not preferred. The flags you see in IINA that convention were likely added by me. I was used to that convention from other projects.

low-batt committed:
Combine separate flags into a player state property

This commit will:
- Add a PlayerState enum that defines the mutually exclusive states a
  PlayerCore can be in
- Add a property to PlaybackInfo to hold a player's state
- Remove the isIdle, isPaused, isPlaying and isSeeking properties from
  PlaybackInfo
- Remove the fileLoading, isStopping, isStopped, isShuttingDown and
  isShutdown properties from PlayerCore
- Remove the fileLoaded property from MPVController
- Move the receivedEndFileWhileLoading property from MPVController to
  PlayerCore
- Replace checking of the stopping and shutting down properties with
  checking a new PlayerState.active computed property
- Replace checking of the fileLoaded property with checking a new
  PlayerState.loaded computed property
- Move remaining MPV_EVENT_SHUTDOWN processing from MPVController to
  PlayerCore and the main thread to avoid the need for locks
- Move MPV_EVENT_END_FILE and  idle-active processing from MPVController
  to PlayerCore
- Add an active guard to PlaylistViewController.reloadData
- Correct checking for shutting down in
  MiniPlayerWindowController.windowWillClose

This makes the state the player is in explicit in one property. This
also corrects some errors in the app termination sequence.

uiryuu commented on iina/PlayerCore.swift:
```diff
@@ -1794,15 +1802,32 @@ class PlayerCore: NSObject {
+    syncUI(.playlist)
+  }
+
+  func fileEnded(doToStopCommand: Bool) {
```
Did you mean `dueToStopCommand`?

low-batt replied:
Good catch! Fixed.

uiryuu reviewed (commented):
The changes look much cleaner now.

low-batt committed:
Combine separate flags into a player state property

This commit will:
- Add a PlayerState enum that defines the mutually exclusive states a
  PlayerCore can be in
- Add a property to PlaybackInfo to hold a player's state
- Remove the isIdle, isPaused, isPlaying and isSeeking properties from
  PlaybackInfo
- Remove the fileLoading, isStopping, isStopped, isShuttingDown and
  isShutdown properties from PlayerCore
- Remove the fileLoaded property from MPVController
- Move the receivedEndFileWhileLoading property from MPVController to
  PlayerCore
- Replace checking of the stopping and shutting down properties with
  checking a new PlayerState.active computed property
- Replace checking of the fileLoaded property with checking a new
  PlayerState.loaded computed property
- Move remaining MPV_EVENT_SHUTDOWN processing from MPVController to
  PlayerCore and the main thread to avoid the need for locks
- Move MPV_EVENT_END_FILE and  idle-active processing from MPVController
  to PlayerCore
- Add an active guard to PlaylistViewController.reloadData
- Correct checking for shutting down in
  MiniPlayerWindowController.windowWillClose

This makes the state the player is in explicit in one property. This
also corrects some errors in the app termination sequence.

lhc70000 commented on iina/MPVController.swift:
```diff
@@ -1109,12 +1105,10 @@ class MPVController: NSObject {
       DispatchQueue.main.async { [self] in
-        player.info.isIdle = false
-        player.info.isSeeking = false
         // When playback is paused the display link may be shutdown in order to not waste energy.
```
No state update after seeking? This doesn't seem to be correct.

low-batt replied:
Good catch. The `isSeeking` flag needs to be kept separate. It can't be included in state enum. I've pushed an update.

low-batt committed:
Combine separate flags into a player state property

This commit will:
- Add a PlayerState enum that defines the mutually exclusive states a
  PlayerCore can be in
- Add a property to PlaybackInfo to hold a player's state
- Remove the isIdle, isPaused and isPlaying properties from PlaybackInfo
- Remove the fileLoading, isStopping, isStopped, isShuttingDown and
  isShutdown properties from PlayerCore
- Remove the fileLoaded property from MPVController
- Move the receivedEndFileWhileLoading property from MPVController to
  PlayerCore
- Replace checking of the stopping and shutting down properties with
  checking a new PlayerState.active computed property
- Replace checking of the fileLoaded property with checking a new
  PlayerState.loaded computed property
- Move remaining MPV_EVENT_SHUTDOWN processing from MPVController to
  PlayerCore and the main thread to avoid the need for locks
- Move MPV_EVENT_END_FILE and  idle-active processing from MPVController
  to PlayerCore
- Add an active guard to PlaylistViewController.reloadData
- Correct checking for shutting down in
  MiniPlayerWindowController.windowWillClose

This makes the state the player is in explicit in one property. This
also corrects some errors in the app termination sequence.
